src/System.Memory/src/System/SpanHelpers.char.cs

```diff
@@ -0,0 +1,271 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+#if !netstandard
+using Internal.Runtime.CompilerServices;
+#endif
+
+#if !netstandard11
+using System.Numerics;
+#endif
+
+namespace System
+{
+    internal static partial class SpanHelpers
+    {
+        private static readonly Vector<UInt16> s_equalityTester = new Vector<UInt16>(UInt16.MaxValue);
```

Comment:
These should not be `readonly` statics - they should be inlined constants in the code. `readonly` statics have worse and unpredictable perf characteristics than inlined constants.

Reply:
What qualifier, if any, should I use on the declaration within the code?

Or should it just be as follows, right before the loop body:
`var equalityTester = new Vector<UInt16>(UInt16.MaxValue);`?

Reply:
> right before the loop body

Pass the buck to the Jit as to whether to hoist? 

For this; putting them before the loop body would want 9 registers (as lifetime wraps the loop) and their aren't that many available:
```csharp
if (Vector.GreaterThanAny<UInt16>(value, s_isLatin1Mask)) break;

if (Vector.GreaterThanAny<UInt16>(value, s_whiteSpace13Mask) ||
    Vector.LessThanAny<UInt16>(value, s_whiteSpace9Mask))
{
    Vector<UInt16> comparison = Vector<UInt16>.Zero;
    comparison |= Vector.Equals(value, s_whiteSpace32Mask);
    comparison |= Vector.Equals(value, s_whiteSpace133Mask);
    comparison |= Vector.Equals(value, s_whiteSpace160Mask);
    if (!s_equalityTester.Equals(comparison)) break;
```
Might want to see how something more like the following performs and what the asm diff is: 
```csharp
if (Vector.GreaterThanAny<UInt16>(value, new Vector<UInt16>(0xFF))) break; // Out of whitespace range

if (Vector.GreaterThanAny<UInt16>(value, new Vector<UInt16>(0xd)) ||       // Carriage return
    Vector.LessThanAny<UInt16>(value, new Vector<UInt16>(0x9)))            // Horizontal tab
{
    Vector<UInt16> comparison = Vector<UInt16>.Zero;
    comparison |= Vector.Equals(value, new Vector<UInt16>(0x20);           // Space
    comparison |= Vector.Equals(value, new Vector<UInt16>(0x85);           // Not reachable due to if condition?
    comparison |= Vector.Equals(value, new Vector<UInt16>(0xa0));          // Not reachable due to if condition?
    if (!Vector.Equals(comparison, Vector<UInt16>.Zero)
                .Equals(Vector<UInt16>.Zero) break;
```
Not sure how the Jit behaves when choosing to hoist vectors; but worth looking into?

Reply:
Though it results in overall smaller asm size for the entire method, it increases the loop body size.

loop body size: 248 vs 310 bytes (after)
total method asm size: 640 vs 453 bytes (after)

It improves perf for span.Length < 32 though, at the cost of larger spans:
![image](https://user-images.githubusercontent.com/6527137/35354978-bbcd42c6-0100-11e8-91f1-dbfa77de3cf9.png)


![image](https://user-images.githubusercontent.com/6527137/35354891-65159ea6-0100-11e8-9fa3-8190e8b427a3.png)

Reply:
I clearly underestimated the number of registers 😄 
![image](https://user-images.githubusercontent.com/1142958/35356516-9728c7e0-0148-11e8-9655-8179afc2472c.png)
Also the Jit hoisting 😢